pkg/proctree/nodes/process.go

```diff
@@ -0,0 +1,239 @@
+package nodes
+
+import "sync"
+
+// Arguments to feed or retrieve a process node
+
+type ProcessInfo struct {
+	Tid       int32
+	NsTid     int32
+	Pid       int32
+	NsPid     int32
+	StartTime uint64
+	ExitTime  uint64
+}
+
+type FileInfo struct {
+	Name      string
+	Path      string
+	Device    uint32
+	Ctime     uint64
+	Inode     uint64
+	InodeMode uint16
+}
+
+type ProcessForkData struct {
+	Parent ProcessInfo
+	Child  ProcessInfo
+}
+
+type ProcessExecData struct {
+	Command     FileInfo
+	Interpreter FileInfo
+	Pid         int32
+	StartTime   uint64
+}
+
+type ProcessExitData struct {
+	ExitedPid int32
+	ExitCode  int64
+	StartTime uint64
+	ExitTime  uint64
+	GroupExit bool
+}
+
+//
+// Process Node
+//
+
+type Process struct {
```

Comment:
Sure, I see your point. This is/was unfinished exactly for this discussion. The naming "process/thread" does not mean anything else but describing a "task" managed by the kernel scheduler. isThread was supposed to be set if the 'task' is/was a LWP. I did not want to "set" this as the final data modeling.

The idea here is/was this (and yes, unfinished, opened to discussions, etc, just a prototype).

1. For each OS task we have a hash (consisting of its TID + start_time)

2. Each task might be:
   - a process (its parent field has the hash of another process)
   - a thread (lwp) (its parent field probably should have the hash of the parent of the thread group leader ?)

3. Each "Process" will have the same concept of time series through an internal pointer:
```go
   type Process struct {
     Versions *TimeSeries // something like this (like you've done already)
   }
```

So everytime anything regarding that process (or thread) changes, we will save a new copied version of that object, indexed by its self per time basis (so you can pick its version by time, just like you had). So, in this case, if we have the following scenario:

1. Process A, Process B and Process C.
2. Process A creates Thread A1, A2.
3. Process B creates Thread B1.
4. Process C creates Process D.

We will have one `struct Process` (yes we can rename to 'task' or something better) per each existing task (being process or thread). The Thread B1, for example, will have its hash (TID + starttime) and its 'parent hash field' will be process B hash. 

Since the hashes are always unique, if the process D, created by process C, is reparanted to Process A, for example, then we will have its version pointing to the hash of Process C, then its last version will have the hash of Process A as parent (together with PPID, and all changes made).

The "hash" here works like a database primary key. You guarantee its uniqueness, data integrity, indexing performance, etc. It allows us to transact changes to the entire tree without holding locks as well (thinking about when keeping files accessed, sockets created, permissions changed, etc... info in here).

Reply:
I agree with you that from the kernel POW, both processes and threads are tasks.
However, as I believe that Tracee (at least in user-mode) is a security tool more that low-level monitoring tool, then the distinction is important. For security events, threads are not interesting most of the times. The most relevant information of a thread is its name in the security point of view. Most of the time, security researchers look at process tree, because the binary or script executed is the main focus.
So, I think that it is important to decide what representation of tasks/processes we want to support.
If you thing you can maintain the process tree as tasks, but expose it to the users as processes and threads, then there is no real issue. Else, I think we should think how to represent it.
I liked here the fact that your process tree is very simple, having a single type of nodes. I think we can keep it, making the tree a thread-tree, but we need the nodes to share the process information (via the time series like you propose sounds good - though the times series object is mutable so I wander how you will support it while keeping it thread-safe).
Anyways, I look forward to your final design :)

Reply:
Another thing I want to emphasize - even though the kernel is changing the parent of the process in some occasions, when a malicious act is detected, this is not interesting.
What is important most of the time is what was the process that forked the current one. Even if it died, and its forking process died as well by the check time. This is because we want to know the history of security events.
So, your current approach of changing the parent is all good if you want to also expose how the kernel see the process tree, but you must keep in all cases the real parent from the fork point of view.

Reply:
> So, your current approach of changing the parent is all good if you want to also expose how the kernel see the process tree, but you must keep in all cases the real parent from the fork point of view.

Yep, I believe this would be the "oldest" version in timeseries (first parent seen for that task, likely from SchedProcessFork). Ok, lets continue and you keep telling me what interests most from the security POV.